member/src/main/scala/com/improving/app/member/domain/Member.scala

```diff
@@ -56,220 +68,409 @@ object Member extends StrictLogging {
+              meta = memberInfoEdited.meta.get
+            )
+          case _: TerminatedMemberState => state
         }
+
+      case other =>
+        throw new RuntimeException(s"Invalid/Unhandled event $other")
     }
+  }
 
-  def activateMember(
-      memberId: MemberId,
-      activatingMember: MemberId,
-      replyTo: ActorRef[StatusReply[MemberResponse]]
+  private def registerMemberLogic(
+    meta: MemberMetaInfo,
+    registerMemberCommand: RegisterMember,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
   ): ReplyEffect[MemberEvent, MemberState] = {
-    val event = MemberActivated(Some(memberId), Some(activatingMember), Some(Timestamp(Instant.now(clock))))
-
-    Effect
-      .persist(event)
-      .thenReply(replyTo) { _ => StatusReply.Success(MemberEventResponse(event)) }
+    if (meta.createdBy.isDefined) {
+      useErrorStatusReply(replyTo, s"Member has already been registered.")
+    } else {
+      val now = Some(Timestamp(Instant.now(clock)))
+      val newMeta = meta.copy(
+        lastModifiedOn = now,
+        lastModifiedBy = registerMemberCommand.registeringMember,
+        createdOn = now,
+        createdBy = registerMemberCommand.registeringMember,
+        currentState = MEMBER_STATUS_DRAFT
+      )
+      val event =
+        MemberRegistered(registerMemberCommand.memberId, registerMemberCommand.memberInfo, Some(newMeta))
+      Effect.persist(event).thenReply(replyTo) { _ =>
+        val res = StatusReply.Success(MemberEventResponse(event))
+        logger.info(s"registerMember: $res")
+        res
+      }
+    }
   }
 
-  def inactivateMember(
-      memberId: MemberId,
-      actingMember: MemberId,
-      replyTo: ActorRef[StatusReply[MemberResponse]]
+  private def registerMember(
+    state: MemberState,
+    registerMemberCommand: RegisterMember,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
   ): ReplyEffect[MemberEvent, MemberState] = {
+    MemberValidation.validateRegisterMember(registerMemberCommand) match {
+      case Valid(_) =>
+        state match {
+          case DraftMemberState(_, _, meta) => registerMemberLogic(meta, registerMemberCommand, replyTo)
+          case _: RegisteredMemberState => useErrorStatusReply(replyTo, s"Member has already been registered.")
+          case _: TerminatedMemberState => useTerminatedStateErrorStatusReply(replyTo)
+        }
+      case Invalid(errors) =>
+        useCommandValidationErrorStatusReply(replyTo, errors, "RegisterMember")
+    }
+  }
 
-    val event = MemberInactivated(Some(memberId), Some(actingMember), Some(Timestamp(Instant.now(clock))))
-    Effect
-      .persist(event)
-      .thenReply(replyTo) { _ => StatusReply.Success(MemberEventResponse(event)) }
+  private def activateMemberLogic(
+    info: MemberInfo,
+    meta: MemberMetaInfo,
+    activateMemberCommand: ActivateMember,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
+  ): ReplyEffect[MemberEvent, MemberState] = {
```

Comment:
It's not clear to me what the separation of concerns is meant to be between the *Logic methods and the calling methods. I found in my Tenant changes that teasing out all the actual message sending to the calling method helped to clarify things quite a bit and decouple the business logic from the mechanical aspect of the actor. There's also a bunch of validation that appears very similar between all these command handlers and maybe could also be teased out into some shared code to help the command handlers focus on the unique logic associated with each one.